draco/analysis/qstack.py

```diff
@@ -0,0 +1,148 @@
+                int(nfreq / 2) - self.freqside : int(nfreq / 2) + self.freqside + 1
+            ]
+        )
+        self.stack_axis["centre"] = (
+            self.stack_axis["centre"] - self.stack_axis["centre"][self.freqside]
+        )
+
+        # Get f_mask and qs_indices
+        freqdiff = freq[np.newaxis, :] - qso_freq[:, np.newaxis]
+        # Stack axis bin edges to digitize each quasar at.
+        stackbins = self.stack_axis["centre"] + 0.5 * self.stack_axis["width"]
+        stackbins = np.append(
+            stackbins,
+            self.stack_axis["centre"][-1] - 0.5 * self.stack_axis["width"][-1],
+        )
+        # Index of each frequency in stack axis, for each quasar
+        qs_indices = np.digitize(freqdiff, stackbins) - 1
+        # Indices to be processed in full frequency axis for each quasar
+        f_mask = (qs_indices >= 0) & (qs_indices < self.nstack)
+        # Only quasars in the frequency range of the data.
+        qso_mask = (np.sum(f_mask, axis=1) > 0).astype(bool)
+
+        # Reduce mask and indices to this process range
+        # to reduce striding through this data
+        qso_mask = qso_mask[loff : loff + lshape]
+        qs_indices = qs_indices[loff : loff + lshape]
+        f_mask = f_mask[loff : loff + lshape]
+
+        # Quasar stack array.
+        quasar_stack = np.zeros(self.nstack, dtype=np.float)
+        quasar_weight = np.zeros(self.nstack, dtype=np.float)
+
+        qcount = 0  # Quasar counter
+        # For each quasar in the range of this process
+        for lq, gq in formed_beam.beam[:].enumerate(axis=0):
+            if not qso_mask[lq]:
+                # Quasar not in the data redshift range
+                continue
+            qcount += 1
+            # Indices and slice for frequencies included in the stack.
+            f_indices = np.arange(nfreq, dtype=np.int32)[f_mask[lq]]
+            f_slice = np.s_[f_indices[0] : f_indices[-1] + 1]
+
+            quasar_stack += np.bincount(
+                qs_indices[lq][f_slice],
+                weights=(
+                    formed_beam.beam[gq, 0][f_slice]
+                    * formed_beam.weight[gq, 0][f_slice]
+                ),
+                minlength=self.nstack,
+            )
+
+            quasar_weight += np.bincount(
+                qs_indices[lq][f_slice],
+                weights=formed_beam.weight[gq, 0][f_slice],
+                minlength=self.nstack,
+            )
+
+            # mpiutil.barrier()  # TODO: I am not sure I need a barrier here.
+        # Gather quasar stack for all ranks. Each contains the sum
+        # over a different subset of quasars.
+        quasar_stack_full = np.zeros(
+            mpiutil.size * self.nstack, dtype=quasar_stack.dtype
+        )
+        quasar_weight_full = np.zeros(
+            mpiutil.size * self.nstack, dtype=quasar_weight.dtype
+        )
+        # Gather all ranks
+        mpiutil.world.Allgather(quasar_stack, quasar_stack_full)
```

Comment:
Try to avoid using `mpiutil` when you can. In this case you can get the communicator out of the input data and just use `comm.Allgather`

Reply:
You may want to use `allgather`, which will involve creating the destination array, but will be faster and scale to large numbers of quasars.

Reply:
Hmm. I am not sure what you mean. I do create the destination arrays (`quasar_stack_full`) and I thought the upper-case versions were the buffer-object, fast ones.
Also, keep in mind that this gathering is done from an object of size `nstack` (number of frequencies in the quasar stack, so around 100) to an object of size `(nprocess * nstack)`, so it's very small...